Remove usage of IndexSearcher#search(Query, Collector) from join package

msfroh opened the pull request:
### Description
Relates to https://github.com/apache/lucene/issues/12892

For global ordinal-based join, we can support concurrent search. For numeric and term-based joins, we fail if we're called from a multithreaded searcher.

I can implement concurrent versions of the other join collectors, but wanted to get a first pass that removes the uses of `IndexSearcher#search(Query, Collector)`.

Note that for the cases that still assume a single-threaded searcher, I used a version of `CollectorManager#wrap(Collector)` from https://github.com/apache/lucene/pull/13735, with my guess for where it will end up based on feedback so far.

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

msfroh committed:
Remove usage of IndexSearcher#search(Query, Collector) from join package

For global ordinal-based join, we can support concurrent search. For
others, we fail if we're called from a multithreaded searcher.

javanna commented on lucene/join/src/java/org/apache/lucene/search/join/JoinUtil.java:
```diff
@@ -530,49 +528,104 @@ public static Query createJoinQuery(
+          LongBitSet collectedOrdinals =
+              searcher.search(
+                  rewrittenFromQuery,
+                  new CollectorManager<GlobalOrdinalsCollector, LongBitSet>() {
```
do we expect users to benefit from this collector manager outside of `JoinUtil`? If so maybe it deserves to be in its own outer class, and include the merge method (now called combine) in it.

msfroh replied:
Good idea!

I've created a new `MergeableCollector` interface that adds a `merge` method to `Collector`. Then we can reuse a `MergeableCollectorManager` to fold all results into a single collector in the `reduce` step. At the very least, I can use it for the two cases in this method (`GlobalOrdinalsCollector` and `GlobalOrdinalsWithScoreCollector`). I should be able to use them once I tackle the terms and numeric collectors.

I've kept them package-private for now, so I can use them throughout the `join` module. I don't know if anyone else wants to implement `reduce` as a pair-wise merge. (Most of the existing `reduce` implementations just do an N-way merge.)

javanna reviewed (commented):
I think it's a good approach. Let's prioritize removing those leftover usages and add the missing collector managers as a next step.

javanna commented:
@msfroh thanks a lot for picking this up!!!

msfroh committed:
Reuse pairwise merging collector logic

Since I plan to implement numeric and term collectors that support
merging all collectors into a single collector, it makes sense to
move MergeableCollectorManager into its own top-level class.

msfroh commented:
@javanna  -- I've managed to get the remaining numeric / terms collectors in the Join module working with multiple search threads.

I can add them to this PR, but the diff is pretty massive. I'm thinking of holding off for another PR, but I'm happy to go either way. 

There is probably value in "atomically" jumping from the current "always single-threaded" mode straight to "everything works with a multithreaded searcher", versus this PR's current state where global ordinal-based joins work with a multithreaded searcher but numeric/term-based joins don't.

Thanks a lot for the review!

javanna reviewed (commented):
I left a couple more comments, I am fine getting this in and focusing on the missing parallel collector managers as a follow-up. Thanks for the work you put into this!

msfroh commented on lucene/join/src/java/org/apache/lucene/search/join/MergeableCollector.java:
```diff
@@ -0,0 +1,26 @@
+
+  /** Folds the results from another collector of the same type into this collector. */
+  void merge(C collector) throws IOException;
+}
```
Makes sense! 

Given that the merge logic is collector-specific, I'm thinking that collectors should expose what they produce after reduction.

This also makes it easier to push the N-way merge into the collectors.

Maybe the resulting CollectorManager might be more reusable.

msfroh replied:
Actually, maybe we're disagreeing? I think you're suggesting that the CollectorManager should handle merging, while I'm suggesting that Collector should have some output and know how to merge multiple outputs.

This would mean that every collector needs a corresponding CollectorManager, or at least a CollectorManager for every output type, right?

msfroh replied:
Hmm... While it may be a bigger change, it feels like we'll we should really have just one CollectorManager that merges results from Collecors based on what the Collectors collect.

msfroh replied:
I can reimplent my follow-up PR to do three separate CollectorManagers that handle the merge logic, I suppose.

That would be more consistent with the existing Lucene code, but I'm not sure that I agree with that approach.

It essentially means that every Collector needs a parallel CollectorManager that knows how to merge that Collector. For something like these join Collectors, it also means that the CollectorManager needs to know how to implement the ScoreMode (min, max, sum, or avg).

javanna replied:
> This would mean that every collector needs a corresponding CollectorManager, or at least a CollectorManager for every output type, right?

I guess it depends, you could also reuse the same collector manager across collectors if they share a sub-type perhaps. But as a general guideline, yes. After all, search goes through collector managers and collector managers are responsible to create collectors and reduce multiple collectors into a single result, whether that be a collector or some more meaningful result class.

I don't have a strong opinion, but I thought that moving your merge method to the collector manager should not be a big change, I may be missing some details perhaps.

msfroh replied:
Yeah... I think you're right. I'll still need an abstraction to describe what the Collector produces, but then either that "thing" will know how to merge themselves, or it can be an anonymous function passed to the CollectorManager.

msfroh replied:
Oh, wait, I'm a moron. 

I should just more or  less turn all of these Collectors into CollectorManagers. Then I'm not duplicating anything.

msfroh committed:
Wrap all collectors in a CollectorManager

This change removes MergeableCollector (and MergeableCollectorManager),
wraps all of the custom Collector classes in their own CollectorManager,
and removes all remaining occurrences of
CollectorManager.forSequentialExecution from the tests.

This also adds all of the other join collectors, bringing the join
module fully into the CollectorManager club.

msfroh commented:
Okay -- I wrapped all of the `Collector`s in `CollectorManagers`, and managed to remove all uses of `CollectorManager.forSequentialExecution`. I also went ahead and added the remaining `Collector`s to this PR.

msfroh commented:
@javanna -- obviously we missed the Lucene 10 cutoff for this, but it doesn't break any public APIs. Do you think it makes sense to merge this as an incremental improvement?

javanna commented:
Hey @msfroh I need to review this once again and resume context on it. I will try to get to it this week.
